src/Xamarin.Android.Build.Tasks/Utilities/MarshalMethodsNativeAssemblyGenerator.cs

```diff
@@ -177,40 +205,86 @@ sealed class MarshalMethodsManagedClass
 		List<MarshalMethodInfo> methods;
 		List<StructureInstance<MarshalMethodsManagedClass>> classes = new List<StructureInstance<MarshalMethodsManagedClass>> ();
 
+		public MarshalMethodsNativeAssemblyGenerator (int numberOfAssembliesInApk, ICollection<string> uniqueAssemblyNames, IDictionary<string, IList<MarshalMethodEntry>> marshalMethods, TaskLoggingHelper logger)
+		{
+			this.numberOfAssembliesInApk = numberOfAssembliesInApk;
+			this.uniqueAssemblyNames = uniqueAssemblyNames ?? throw new ArgumentNullException (nameof (uniqueAssemblyNames));
+			this.marshalMethods = marshalMethods;
+			this.logger = logger ?? throw new ArgumentNullException (nameof (logger));
+
+			if (uniqueAssemblyNames.Count != numberOfAssembliesInApk) {
+				throw new InvalidOperationException ("Internal error: number of assemblies in the apk doesn't match the number of unique assembly names");
+			}
+		}
+
 		public override void Init ()
 		{
-			Console.WriteLine ($"Marshal methods count: {MarshalMethods?.Count ?? 0}");
-			if (MarshalMethods == null || MarshalMethods.Count == 0) {
+			Console.WriteLine ($"Marshal methods count: {marshalMethods?.Count ?? 0}");
+			if (marshalMethods == null || marshalMethods.Count == 0) {
 				return;
 			}
 
 			var seenClasses = new Dictionary<string, int> (StringComparer.Ordinal);
-			methods = new List<MarshalMethodInfo> ();
-			foreach (IList<MarshalMethodEntry> entryList in MarshalMethods.Values) {
+			var allMethods = new List<MarshalMethodInfo> ();
+
+			// It's possible that several otherwise different methods (from different classes, but with the same
```

Comment:
Is there an example of this?  Given that the `Java_…` names contain the declaring package name and declaring type name, *how* can two different methods "from different classes" have the same name?

Or is "name" in this context the *method* name, not the `Java_…` name?

Reply:
Yep, there are a few examples in the Hello Maui sample app:

```
Overloaded MM: Java_crc64e1fb321c08285b90_CellAdapter_n_1onActionItemClicked
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean Android.Views.ActionMode/ICallback::OnActionItemClicked(Android.Views.ActionMode,Android.Views.IMenuItem))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onActionItemClicked'
  Using FULL signature
    mangling 'Landroid/view/ActionMode;Landroid/view/MenuItem;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onActionItemClicked__Landroid_view_ActionMode_2Landroid_view_MenuItem_2
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean AndroidX.AppCompat.View.ActionMode/ICallback::OnActionItemClicked(AndroidX.AppCompat.View.ActionMode,Android.Views.IMenuItem))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onActionItemClicked'
  Using FULL signature
    mangling 'Landroidx/appcompat/view/ActionMode;Landroid/view/MenuItem;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onActionItemClicked__Landroidx_appcompat_view_ActionMode_2Landroid_view_MenuItem_2

Overloaded MM: Java_crc64e1fb321c08285b90_CellAdapter_n_1onCreateActionMode
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean Android.Views.ActionMode/ICallback::OnCreateActionMode(Android.Views.ActionMode,Android.Views.IMenu))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onCreateActionMode'
  Using FULL signature
    mangling 'Landroid/view/ActionMode;Landroid/view/Menu;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onCreateActionMode__Landroid_view_ActionMode_2Landroid_view_Menu_2
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean AndroidX.AppCompat.View.ActionMode/ICallback::OnCreateActionMode(AndroidX.AppCompat.View.ActionMode,Android.Views.IMenu))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onCreateActionMode'
  Using FULL signature
    mangling 'Landroidx/appcompat/view/ActionMode;Landroid/view/Menu;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onCreateActionMode__Landroidx_appcompat_view_ActionMode_2Landroid_view_Menu_2

Overloaded MM: Java_crc64e1fb321c08285b90_CellAdapter_n_1onDestroyActionMode
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Void Android.Views.ActionMode/ICallback::OnDestroyActionMode(Android.Views.ActionMode))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onDestroyActionMode'
  Using FULL signature
    mangling 'Landroid/view/ActionMode;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onDestroyActionMode__Landroid_view_ActionMode_2
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Void AndroidX.AppCompat.View.ActionMode/ICallback::OnDestroyActionMode(AndroidX.AppCompat.View.ActionMode))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onDestroyActionMode'
  Using FULL signature
    mangling 'Landroidx/appcompat/view/ActionMode;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onDestroyActionMode__Landroidx_appcompat_view_ActionMode_2

Overloaded MM: Java_crc64e1fb321c08285b90_CellAdapter_n_1onPrepareActionMode
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean Android.Views.ActionMode/ICallback::OnPrepareActionMode(Android.Views.ActionMode,Android.Views.IMenu))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onPrepareActionMode'
  Using FULL signature
    mangling 'Landroid/view/ActionMode;Landroid/view/Menu;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onPrepareActionMode__Landroid_view_ActionMode_2Landroid_view_Menu_2
  implemented in: Microsoft.Maui.Controls.Handlers.Compatibility.CellAdapter (System.Boolean AndroidX.AppCompat.View.ActionMode/ICallback::OnPrepareActionMode(AndroidX.AppCompat.View.ActionMode,Android.Views.IMenu))
    mangling 'crc64e1fb321c08285b90/CellAdapter'
    mangling 'n_onPrepareActionMode'
  Using FULL signature
    mangling 'Landroidx/appcompat/view/ActionMode;Landroid/view/Menu;'
     new native symbol name: Java_crc64e1fb321c08285b90_CellAdapter_n_1onPrepareActionMode__Landroidx_appcompat_view_ActionMode_2Landroid_view_Menu_2
```

Reply:
Either you updated the comment, or I *horribly* misread this comment; I thought you were saying that the *long*/*full* name had conflicts/collisions.

What you're showing here are *overloads*.

For example, consider `Activity.startActivity()`:

  * [`Activity.startActivity(Intent)`](https://developer.android.com/reference/android/app/Activity#startActivity(android.content.Intent))
  * [`Activity.startActivity(Intent, Bundle)`](https://developer.android.com/reference/android/app/Activity#startActivity(android.content.Intent,%20android.os.Bundle))

You should be able to override *both* of these in C#:

```csharp
class MyActivity : Activity {
    public override unsafe void StartActivity (Android.Content.Intent? intent) => …;
    public override unsafe void StartActivity (Android.Content.Intent? intent, Android.OS.Bundle? options) => …;
}
```

This would also result in needing to use the full `Java_…` name.  (This would likely also make a good unit test in a future commit...)

Updating the comment to use the term "method overload" would be a helpful clarification.  The fact that the methods being overridden came from different classes/interfaces is only slightly relevant; the more important bit is that within the JCW type, these are method *overloads*, a'la:

```java
interface IntFooable {
    void foo(int value);
}
interface StringFooable {
    void foo(String value);
}
class MyFooable implements IntFooable, StringFooable {
    public native void foo(int value) {}
    public native void foo(String value) {}
}
```

Here, you can't just have `Java_example_MyFooable_foo`, because the method is overloaded.  Instead, the full names must be used.

Reply:
I modified the comment by adding the examples, but the rest of the wording remained the same. The "simple" overloads you give above are handled in the classifer, the case handled here is when two different managed types will end up having the same Java package name (like the "standard" Android vs AndroidX overloads in the comment) and thus become "native" overloads, requiring native symbol name with full signature.